start: Add new experiments section (#4393)
iterative/dvc.org: content/docs/start/experiments/experiment-versioning.md

---
title: 'Get Started: Experiment Versioning'
description:
  'Get started with versioning experiments with DVC. Track the changes to the
  code, data, metrics, parameters and plots associated with each experiment,
  without bloating your Git repo.'
---

# Get Started: Experiment Versioning

> The early stages of a Machine Learning (ML) project are usually exploratory and
> demand fast iterations. Many ML practitioners chose to use tools that require
> little setup and enable rapid prototyping, like those based on
> [Jupyter Notebooks](https://jupyter.org/).
> 
> Although these tools are extremely useful, keeping track of the changes you made
> across experiments quickly becomes cumbersome and reproducing an experiment can
> be a challenge.
> 
> You can start using DVC to version your experiments without leaving your Jupyter
> Notebook. There are no logins, servers, databases, or UI to spin up. Every
> <abbr>DVC experiment</abbr> will be versioned **without cluttering your repo**,
> unlike saving each run to a separate directory or creating a Git branch for
> each.

Reviewer: ```suggestion
Tools like [Jupyter Notebooks](https://jupyter.org/) are useful for rapid prototyping,
but it's hard to keep track of changes and reproduce experiments.
You can start using DVC to version your experiments without leaving your Jupyter
Notebook. There are no logins, servers, databases, or UI to spin up. Every
<abbr>DVC experiment</abbr> will be versioned **without cluttering your repo**,
unlike saving each run to a separate directory or creating a Git branch for
each.
```

Missed this the 1st time, but let's try to get this down to 1 short paragraph if possible. Comparing to other tools, they don't even give any problem statement like this, so not sure the first sentence of the suggestion is even needed.

...

trainer.add_callback(DVCLiveCallback(save_dvc_exp=True))
trainer.train()
```

</tab>

<tab title="Keras">

```python
from dvclive.keras import DVCLiveCallback

...

model.fit(
  train_dataset, validation_data=validation_dataset,
  callbacks=[DVCLiveCallback(save_dvc_exp=True)])
```

</tab>

<tab title="General Python API">

```python
from dvclive import Live

with Live(save_dvc_exp=True) as live:
    live.log_param("epochs", NUM_EPOCHS)

    for epoch in range(NUM_EPOCHS):
        train_model(...)
        metrics = evaluate_model(...)
        for metric_name, value in metrics.items():
            live.log_metric(metric_name, value)
        live.next_step()
```

</tab>

</toggle>

After this, each execution of the code will create a <abbr>DVC experiment</abbr>
containing the results and the changes needed to reproduce it.

[DVCLive] will automatically log some metrics, parameters and plots from the ML
Framework but you can also manually
[log additional info](/doc/dvclive/api-reference#log-data) to be included in the
experiment.

If you have [data tracked by DVC](/doc/start/data-management/data-versioning),
it will be automatically included in the experiment.

> Once you have added [DVCLive] to your code, you can track the changes with git:

Reviewer: Not sure I get what this sentence is trying to say. Are we trying to make sure the code gets included with the experiments? If so, maybe it needs to be more explicit, like `your code will be included with your experiment if the code file is tracked by git` or something like that.

Author: I am removing it for now and revisit it if/when we try to make the example more reproducible or tied to the example repo.

---
title: 'Get Started: Experiment Versioning'
description:
  'Get started with versioning experiments with DVC. Track the changes to the
  code, data, metrics, parameters and plots associated with each experiment,
  without bloating your Git repo.'
---

> # Get Started: Experiment Versioning

Reviewer: I think the intro should at least mention that experiments are saved with git (I like to say "as hidden git commits"). Reasons:

1. Without this, some of the rest of the doc becomes a little confusing (like when we say to make sure you git add your code).
2. A big part of the value prop is that you manage both your code and experiment info in one place. We can't get into a list of benefits of our approach, but this one seems important enough to highlight.

Author: I think that, on the first page, we should focus on getting people to see stuff in the UIs as soon as possible, not distracting with other details, to match the effort to be onboarded in other tools. 

We do have some hints that something different is going on using the on-hover details that you improved and some mentions to Git. 
I think that should be all for this page

> A big part of the value prop is that you manage both your code and experiment info in one place. We can't get into a list of benefits of our approach, but this one seems important enough to highlight.

I think that is where the second page comes in. Quoting from your other comment:

> I'm not sure what to do with the experiment management page. It already feels like it goes pretty deep, certainly beyond where other experiment tracking tools go (since they don't really even cover this topic and some of it's not applicable to them). Should we be covering it in get started at all? 

Once people can see stuff in the UIs, I think it is good to have a separate page where we go into the value proposition against other tools. 
I agree we should try to keep it as lightweight as possible.

> Which commands/features are most important?

I think we need the 3 existing points `sharing`, `persisting`, and `cleaning`. Not covering one of those points would feel like an incomplete workflow to me.

Maybe we could try to reduce complexity by being more opinionated (i.e. don't mention `apply` but only `branch`)

Reviewer: Okay, with the updated tooltip, it seems less problematic that we don't explain in the page itself.

> Once people can see stuff in the UIs, I think it is good to have a separate page where we go into the value proposition against other tools.

Do you think the 2nd page conveys that value prop?

Reviewer: Okay, I think it's good enough for now. Resolving this one.

The early stages of a Machine Learning (ML) project are usually exploratory and
demand fast iterations. Many ML practitioners chose to use tools that require
little setup and enable rapid prototyping, like those based on
[Jupyter Notebooks](https://jupyter.org/).

Although these tools are extremely useful, keeping track of the changes you made
across experiments quickly becomes cumbersome and reproducing an experiment can
be a challenge.

You can start using DVC to version your experiments without leaving your Jupyter
Notebook. There are no logins, servers, databases, or UI to spin up. Every
experiment will be versioned **without cluttering your repo**, unlike saving
each run to a separate directory or creating a Git branch for each.

## Enable Versioning

All you need to start is a <abbr>DVC repository</abbr> and the [DVCLive] Python
library installed:

```cli
$ pip install dvclive
```

> In your Python code, you can start versioning your experiments with
> `save_dvc_exp=True`.

Reviewer: ```suggestion
In your Python code, you can start versioning your experiments in
DVCLive's `Live` API or framework-specific callbacks with `save_dvc_exp=True`.
```

Minor, not blocking suggestion to stress the overall API here since this is not a DVCLive-specific page and that context may not be so obvious.

There are some examples below
([other frameworks available](/doc/dvclive/api-reference/ml-frameworks)):

<toggle>

<tab title="Pytorch Lightning">

```python
from dvclive.lightning import DVCLiveLogger

...

trainer = Trainer(logger=DVCLiveLogger(save_dvc_exp=True))
trainer.fit(model)
```

</tab>

<tab title="Hugging Face">

```python
from dvclive.huggingface import DVCLiveCallback

...

trainer.add_callback(DVCLiveCallback(save_dvc_exp=True))

> trainer.train()

Author: > should we use the context manager everywhere to be safe?

Not sure.

Given that we are showing the basic callback usage, I am not sure it's worth it to use the context manager here. 
Just including it without practical use would make the basic usage look more complicated than it is.

We could show logging additional stuff after the callback but that would complicate the overall example

```

</tab>

<tab title="Keras">

```python
from dvclive.keras import DVCLiveCallback

...

model.fit(
  train_dataset, validation_data=validation_dataset,
  callbacks=[DVCLiveCallback(save_dvc_exp=True)])
```

</tab>

<tab title="General Python API">

```python
from dvclive import Live

with Live(save_dvc_exp=True) as live:
    live.log_param("epochs", NUM_EPOCHS)

    for epoch in range(NUM_EPOCHS):
        train_model(...)
        metrics = evaluate_model(...)
        for metric_name, value in metrics.items():
            live.log_metric(metric_name, value)
        live.next_step()
```

</tab>

</toggle>

After this, each execution of the code will create a DVC <abbr>experiment</abbr>

> containing the results and the changes needed to reproduce it.

Reviewer: Related to the [comment above](https://github.com/iterative/dvc.org/pull/4393/files#r1139207329), this would be a good alternative place to introduce that experiments are saved with git.

[DVCLive] will automatically log some metrics, parameters and plots from the ML
Framework but you can also manually
[log additional info](/doc/dvclive/api-reference#log-data) to be included in the
experiment.

> If you have [data tracked by DVC](/doc/start/data/data-versioning), it will be

Reviewer: ```suggestion
If you have [data tracked by DVC](/doc/start/data/data-management/data-versioning), it will be
```

automatically included in the experiment.

<admon type="tip">

Make sure that the experiment's code is tracked by Git:

```cli
git add notebooks/TrainSegModel.ipynb
git commit -m "Add Notebook using DVCLive"
```

</admon>

## Monitor the execution

By following the steps above, you enable different options to monitor the
training progress:

<toggle>

<tab title="DVCLive Report">

By default, DVCLive will
[generate or update a report](/doc/dvclive/api-reference/live/make_report)
displaying all the logged data.

If you pass `report="notebook"` to DVCLive, the report will be displayed and
updated inside the output of the cell:

![Notebook report](/img/dvclive-notebook.gif)

</tab>

<tab title="VSCode Extension">

The
[DVC Extension for VS Code](https://marketplace.visualstudio.com/items?itemName=Iterative.dvc)
will also display all the data logged by DVCLive:

![VS Code Report](/img/dvclive-vscode-monitoring.gif)

</tab>

<tab title="Studio">

> If you have followed the instructions in
> [Studio Live Experiments](/doc/studio/user-guide/projects-and-experiments/live-metrics-and-plots),
> DVCLive updates will be displayed in the Studio UI:

Reviewer: Not a blocker: consider giving some motivation here, like:

```suggestion
If you want to share live updates with others or monitor while away from your machine,
follow the instructions in
[Studio Live Experiments](/doc/studio/user-guide/projects-and-experiments/live-metrics-and-plots)
to display updates in the Studio web interface:
```

![Studio Report](/img/dvclive-studio.gif)

</tab>

</toggle>

## Compare results

> After you have executed the code multiple times, you can compare the results:

Reviewer: ```suggestion
After you have run multiple experiments, you can compare the results:
```

💅 Not blocking
